packages/block-editor/src/components/block-edit/index.js

```diff
@@ -10,39 +15,28 @@ import Edit from './edit';
 import { BlockEditContextProvider } from './context';
 
 class BlockEdit extends Component {
-	constructor( props ) {
-		super( props );
-
-		this.setFocusedElement = this.setFocusedElement.bind( this );
-
-		this.state = {
-			focusedElement: null,
-			setFocusedElement: this.setFocusedElement,
-		};
+	constructor() {
+		super( ...arguments );
+
+		// It is important to return the same object if props haven't changed
+		// to avoid  unnecessary rerenders.
+		// See https://reactjs.org/docs/context.html#caveats.
+		this.propsToContext = memize(
+			this.propsToContext.bind( this ),
+			{ maxSize: 1 }
+		);
 	}
 
-	setFocusedElement( focusedElement ) {
-		this.setState( ( prevState ) => {
-			if ( prevState.focusedElement === focusedElement ) {
-				return null;
-			}
-			return { focusedElement };
-		} );
-	}
-
-	static getDerivedStateFromProps( props ) {
-		const { clientId, name, isSelected } = props;
-
-		return {
-			name,
-			isSelected,
-			clientId,
-		};
+	propsToContext( name, isSelected, clientId ) {
+		return { name, isSelected, clientId };
 	}
 
 	render() {
+		const { name, isSelected, clientId } = this.props;
+		const value = this.propsToContext( name, isSelected, clientId );
```

Comment:
This made me think, whether we can use [useMemo](https://reactjs.org/docs/hooks-reference.html#usememo) hook from React. We could convert this component to function. 

> In the future, React may choose to “forget” some previously memoized values and recalculate them on next render, e.g. to free memory for offscreen components. Write your code so that it still works without useMemo — and then add it to optimize performance.

Well, it seems like they will be making it more optimized in the future. Let me know if you want to experiment with it or should I test in the current shape.

Reply:
Would love to do that if there was an option to only cache the last result (cache size of one). I don't think it's a good idea to cache anything more in the context of this render.

Reply:
I found this in React newsletter sent out yesterday: https://github.com/alexreardon/use-memo-one

Let's fly with the current proposal. I guess we can implement our own `useMemize` hook :D

Reply:
I like `memoize-one`. Makes me wish we had something like `memize-one` for purposes like this. It's very useful together with React. I know that you can just pass the option, but I think it's good to provide a shortcut and to discourage `memize` without a limit.  Cc @aduth.

Reply:
Funny that you shared it as I sent a direct message to @aduth asking what he thinks about we introduce `@wordpres/memo` package which would contain:
- `memo`
- `useMemo`

I noticed that `memize` is used with 8 packages as of today and it's not marked as external so we bundle it every time it is used.